### PR TITLE
Remove deepcopy in TensileCreateLibrary

### PR DESCRIPTION
**Pull request by ellosel**
The `deepcopy` function is used throughout Tensile and one of the primary bottlenecks. This PR removes the deep copies of Solutions - namely those that occur in `TensileCreateLibrary`.  This significantly improves the turnaround time of `TensileCreateLibrary` by up to 40% depending on the input to the main.

Ideally, we could make the Solution class immutable as a part of these changes (which we attempted) but the solution changes in ways during runtime that would require a major refactor. 

To ensure that the program still behaves correctly, we ran the following tests:

- Standard CI tests
- A/B comparison to devel of manifest generation using TensileCreateLibrary 
- A/B comparison to devel of manifest generation   using rocBLAS install.sh
- Ran rocBLAS-tests on the feature branch without failure

### CODE REVIEW

File: Tensile/SolutionStructs.py [hunk 1]
```diff
@@ -1720,60 +1721,61 @@ class Solution(collections.abc.Mapping):
...
   ########################################
   # get a list of kernel parameters for this solution
   def getKernels(self):
-    kernel = deepcopy(self)
-    kernel._state.update({"Kernel": True})
+    kernel = self
+    kernel._mutable_state.update({"Kernel": True})
     kernels = []
     kernels.append(kernel)
     return kernels
```
Comment (bstefanuk) [inline Tensile/SolutionStructs.py:1781]: This function will now return a reference to the object itself instead of a full copy, which could have side-effects if calling code requires the original Solution object unchanged. This isn't necessarily avoidable, but it would be great if we could consume the original object.

File: Tensile/SolutionStructs.py [hunk 2]
```diff
@@ -4945,20 +4947,27 @@ def getParameterValueAbbreviation( key, value ):
...
   def __getitem__(self, key):
+    if key in self._mutable_state:
+      return self._mutable_state[key]
     return self._state[key]
 
   def __setitem__(self, key, value):
+    printWarning(f"Mutating solution: {self._name} {key}")
     self._name = None
-    self._state[key] = value
+    self._mutable_state[key] = value
```
Comment (bstefanuk) [inline Tensile/SolutionStructs.py:4971]: What if a key that already exists on `_state` is added via __set_item__? Should we reject over-writing immutable state with mutable state?

Similarly, if a key exists in both `_state` and `_mutable_state`, __getitem__ will prefer the mutable option. Is this the behaviour we're after in all cases?

File: Tensile/TensileCreateLibrary.py [hunk 3]
```diff
@@ -1373,36 +1372,6 @@ def writeBenchmarkClientFiles(libraryWorkingPath, tensileSourcePath, solutions,
...
-    metaData = {"ProblemType": problemType}
-    LibraryIO.writeYAML(mataDataFilePath, metaData)
-
-    codeObjectFiles, newLibrary = writeBenchmarkClientFiles(
-        libraryWorkingPath, tensileSourcePath, solutionList, cxxCompiler
-    )
-
-    return (codeObjectFiles, newLibrary)
-
-
```
Comment (bstefanuk) [inline Tensile/TensileCreateLibrary.py:1405]: Why is this deleted? No longer needed?

**Review by bstefanuk — COMMENTED** [review]
Changes look promising, a few questions about default behaviour and potential side-effects.

File: Tensile/SolutionStructs.py (hunk 2 above)
Comment (ellosel) [reply Tensile/SolutionStructs.py:4971]: The mutated state should be considered the "truth" at least that is my intention. Probably need to clean up a few of the functions to be sure that this is enforced.

File: Tensile/TensileCreateLibrary.py (hunk 3 above)
Comment (ellosel) [reply Tensile/TensileCreateLibrary.py:1405]: I was planning to remove the deepcopy but decided to look to see if this function was used anywhere and I cannot find evidence of it being used.

File: Tensile/SolutionStructs.py (hunk 2 above)
Comment (ellosel) [reply Tensile/SolutionStructs.py]: This impl was abandoned

**Review by babakpst — DISMISSED** [review]
on hold until after FC.

**Comment by nakajee** [comment]
How did you verify this works same as before?

**Comment by ellosel** [comment]
> How did you verify this works same as before?

- Standard CI tests
- A/B comparison to devel of manifest generation using TensileCreateLibrary
- A/B comparison to devel of manifest generation using rocBLAS install.sh
- Ran rocBLAS-tests on the feature branch without failure
- Ran rocblas PTS tests

**Review by nakajee — APPROVED** [review]
Thanks for your comment.
I do not have further comment as long as you confirmed TensileCreateLibrary and rocblas build.
